src/core/lib/view-state.js

```diff
@@ -0,0 +1,51 @@
+export default class ViewState {
+  constructor(props, constraints, userData = {}) {
+    this.props = props;
+    this.constraints = constraints;
+    this.userData = userData;
+    Object.freeze(this);
+  }
+
+  toString(opts) {
+    return this._formatString(opts);
+  }
+
+  // Returns same object if nothing changes (after clamping)
+  getUpdatedState(props) {
+    const newViewState = new ViewState(
+      Object.assign({}, this.props, props),
+      this.constraints,
+      this.userData
+    );
+
+    const viewStateChanged = Object.keys(props).some(
+      key => this.props[key] !== newViewState.props[key]
```

Comment:
This is essentially comparing `props` with `this.props`. Why construct a new `ViewState` object?